chia/wallet/trading/offer.py

```diff
@@ -0,0 +1,385 @@
+            # Check if the puzzle_hash matches the hypothetical `settlement_payments` puzzle hash
+            if addition.puzzle_hash == offer_ph:
+                if tail_hash in offered_coins:
+                    offered_coins[tail_hash].append(addition)
+                else:
+                    offered_coins[tail_hash] = [addition]
+
+        return offered_coins
+
+    def get_offered_amounts(self) -> Dict[Optional[bytes32], int]:
+        offered_coins: Dict[bytes32, List[Coin]] = self.get_offered_coins()
+        offered_amounts: Dict[Optional[bytes32], int] = {}
+        for asset_id, coins in offered_coins.items():
+            offered_amounts[asset_id] = uint64(sum([c.amount for c in coins]))
+        return offered_amounts
+
+    def get_requested_payments(self) -> Dict[bytes32, List[NotarizedPayment]]:
+        return self.requested_payments
+
+    def get_requested_amounts(self) -> Dict[Optional[bytes32], int]:
+        requested_amounts: Dict[Optional[bytes32], int] = {}
+        for asset_id, coins in self.get_requested_payments().items():
+            requested_amounts[asset_id] = uint64(sum([c.amount for c in coins]))
+        return requested_amounts
+
+    def arbitrage(self) -> Dict[Optional[bytes32], int]:
+        offered_amounts: Dict[Optional[bytes32], int] = self.get_offered_amounts()
+        requested_amounts: Dict[Optional[bytes32], int] = self.get_requested_amounts()
+
+        arbitrage_dict: Dict[Optional[bytes32], int] = {}
+        for asset_id in [*requested_amounts.keys(), *offered_amounts.keys()]:
+            arbitrage_dict[asset_id] = offered_amounts.get(asset_id, 0) - requested_amounts.get(asset_id, 0)
+
+        return arbitrage_dict
+
+    # This is a method mostly for the UI that creates a JSON summary of the offer
+    def summary(self) -> Tuple[Dict[str, int], Dict[str, int]]:
+        offered_amounts: Dict[Optional[bytes32], int] = self.get_offered_amounts()
+        requested_amounts: Dict[Optional[bytes32], int] = self.get_requested_amounts()
+
+        def keys_to_strings(dic: Dict[Optional[bytes32], int]) -> Dict[str, int]:
+            new_dic: Dict[str, int] = {}
+            for key in dic:
+                if key is None:
+                    new_dic["xch"] = dic[key]
+                else:
+                    new_dic[key.hex()] = dic[key]
+            return new_dic
+
+        return keys_to_strings(offered_amounts), keys_to_strings(requested_amounts)
+
+    # Also mostly for the UI, returns a dictionary of assets and how much of them is pended for this offer
+    # This method is also imperfect for sufficiently complex spends
+    def get_pending_amounts(self) -> Dict[str, int]:
+        offered_coins: Dict[Optional[bytes32], List[Coin]] = self.get_offered_coins()
+        all_additions: List[Coin] = self.bundle.additions()
+        all_removals: List[Coin] = self.bundle.removals()
+        non_ephemeral_removals: List[Coin] = list(filter(lambda c: c not in all_additions, all_removals))
```

Comment:
I'm not quite sure yet but this and below feel like maybe they should be set operations?  Fix stuff to be sets above and `non_ephemeral_removals = all_removals - all_additions`?  Make it `O(n)` or whatever rather than `O(n*m)`.